src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js

```diff
@@ -1241,4 +1241,53 @@ describe('ReactCompositeComponent', function() {
 
   });
 
+  it('should not overwrite mutated properties', function() {
+    function Moo(props) {
+      React.Component.call(this, props);
+    }
+    Moo.prototype = Object.create(React.Component.prototype, {
+      constructor: {
+        value: Moo,
+      },
+    });
+    Moo.prototype.render = function() {
+      return <span />;
+    };
+    Moo.defaultProps = { idx: 'moo' };
+
+    function Foo(props) {
+      var _props = { idx: this.constructor.defaultProps.idx + '?' };
+
+      React.Component.call(this, _props);
+    }
+    Foo.prototype = Object.create(React.Component.prototype, {
+      constructor: {
+        value: Foo,
+      },
+    });
+    Foo.prototype.render = function() {
+      return <span />;
+    };
+    Foo.defaultProps = { idx: 'foo' };
+
+    function Bar() {
+      Foo.call(this);
+    }
+
+    Bar.prototype = Object.create(Foo.prototype, {
+      constructor: {
+        value: Bar,
+      },
+    });
+    Bar.defaultProps = { idx: 'bar' };
```

Comment:
Yeah, I agree, that would be better.  @drdelambre Would you mind updating this diff?